crypto/arm64cpuid.pl

```diff
@@ -151,7 +151,63 @@
 	lsr	w0,w0,#31
 	ret
 .size	CRYPTO_memcmp,.-CRYPTO_memcmp
+
+.globl	_armv8_rng_probe
+.type	_armv8_rng_probe,%function
+_armv8_rng_probe:
+	mrs	x0, s3_3_c2_c4_0	// rndr
+	mrs	x0, s3_3_c2_c4_1	// rndrrs
+	ret
+.size	_armv8_rng_probe,.-_armv8_rng_probe
+___
+
+sub gen_random {
+my $rdop = shift;
+my $rand_reg = $rdop eq "rndr" ? "s3_3_c2_c4_0" : "s3_3_c2_c4_1";
+
+print<<___;
+.globl	OPENSSL_${rdop}_asm
+.type	OPENSSL_${rdop}_asm,%function
+.align	4
+OPENSSL_${rdop}_asm:
+	mov	x2,xzr
+	mov	x3,xzr
+
+.align	4
+.Loop_${rdop}:
+	cmp	x1,#0
+	b.eq	.${rdop}_done
+	mov	x3,xzr
+	mrs	x3,$rand_reg
+	b.eq	.${rdop}_done
```

Comment:
Isnt zero a possible valid value here?

The documentation for
https://developer.arm.com/documentation/ddi0595/2021-03/AArch64-Registers/RNDR--Random-Number?lang=en

says this

If the hardware returns a genuine random number, PSTATE.NZCV is set to 0b0000.
If the instruction cannot return a genuine random number in a reasonable period of time, PSTATE.NZCV is set to 0b0100 and the data value returned is 0.

Should it be doing
mrs  x4, nzcv

Reply:
Yes, a zero value is possible in the x3 register, though only in a failure situation. 

The `mrs x3, nzcv` is used to clear out the register used for rng (which is not strictly necessary and used to ensure previous values were not in the register) prior to running the command for rng (`mrs x3,$rand_reg`). The check for the failure is done by `b.eq .${rdop}_done` which checks if the Zero flag is set. As quoted from the documentation "If the instruction cannot return a genuine random number in a reasonable period of time, PSTATE.NZCV is set to 0b0100". So the check for the Zero flag would be enough to validate that rng has occurred.

From my understanding of the documentation there is no need to run `mrs x4, nzcv`, prior to rng.